Allow Selection of Flood and Polarization Efficiencies Workspace by File Path

cailafinn opened the pull request:
**Description of work.**
Added text entry boxes to allow specification of a filepath for the polarisation efficiencies workspace and the flood corrections workspace. This will speed up the workflow by removing the need to load the file separately, and allows the information to be held in the batch. Making it a more complete saving of the GUI state. 

**To test:**
1. Talk to Conor or Rachel about getting the corrections files. It's not public, so it can't be put here. 
3. Boot up the Refl GUI
4. Switch the instrument to POLREF.
5. Go to the Experiment Settings Tab
6. Enable polarisation corrections 
7. Check the combo box gets enabled on workspace and becomes a text edit on filepath. 
9. Enter the file path to the downloaded workspace
11. Go back to the Runs tab
12. Enter `31835` into the Run column (not the group, the second box down) and `1.0` into the angle column
13. Click process
14. Check that the `IvsLam_31835` workspace group contains a workspace ending with `_--` and `_++`
15. Do similar with the flood workspaces.

Fixes #34720 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Add option to specify flood workspace by file

RE #34720

Commit:
Add File polarization correction type to model

RE #34720

Commit:
Swtich checkbox for combobox widget

RE #34720

Commit:
Add polarization mode changing methods to view

RE #34720

Commit:
Add path entry widget to experiment view

RE #34720

Commit:
Select correct workspace selection mode

RE #34720

Commit:
Add getters and setters for pol eff widgets

RE #34720

Commit:
Add parameterfile option to pol corr options

I think with the added complexity it's better to be a bit more explicit
here. Rather than just relying on a blank box.

RE #34720

Commit:
Allow explicit selection of parameter file

RE #34720

Commit:
Add options to flood correction combobox

RE #34720

Commit:
Fix encoder and decoder to account for member vars

RE #34720

Commit:
Keep both selector and lineedit in memory

RE #34720

Commit:
Add getters and setters for flood filepath widget

RE #34720

Commit:
Correctly replace ui file flood selector

RE #34720

Commit:
Add mode changing methods to view

RE #34720

Commit:
Disable inputs for none and parameterfile

RE #34720

Commit:
Allow selection of flood workspaces by file path

RE #34720

Commit:
Fix typo in key

RE #34720

Commit:
Fix versioning of batch files

RE #34720

Commit:
Add a release note

RE #34720

Commit:
Fix missed coder tests

RE #34720

rbauststfc reviewed (commented):
The relevant enabling and disabling of the combo boxes is working well, and switching display of the workspace selector and line edit keeps the GUI nice and clean. Loading from a file path or name is generally working well and the corrections are applied as expected. 👍

There were just a few things I noticed while testing.

Flood correction:
- [x] Choosing None from the drop-down and then running a reduction causes an error. My feeling is that it would be good if this was consistent with the way the Polarization Efficiency drop-down works, where selecting None means the correction is not applied.

Polarization efficiencies:
- [x] If I select "Workspace" or "Filepath" from the drop-down but then leave the workspace/filename empty I get error "Correction method is not defined" when running the reduction. When I do this for the flood correction, the correction is simply not applied - it feels like it would be good for the polarization correction to work the same way for consistency.

Both:
- [x] If not too tricky, it might be good to validate the filepath inputs in the GUI, so that the user can see immediately if they have entered an invalid filepath for flood and polarization efficiencies. There is an example of this for the `CalibrationFile` input on the Instrument Settings tab, although I appreciate this is a different input type and so may not be directly relevant here.
- [x] I noticed that when I changed instrument on the Runs tab, clearing of the workspace/filename inputs was a bit inconsistent. For flood correction, any selected workspace is cleared but the filepath input is not. For polarization efficiencies, neither workspace or filepath input are cleared. It would be good to handle this consistently between the two, and I think all should probably revert to default/be cleared on a change in instrument?
- [x] This is a minor thing and may just be my lack of understanding of how things are intended to work, but when I tried to load using a file name it required me to include the file extension for `.nxs` files but not for `.nx5` files. Is this expected?

ConorMFinn commented on qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentPresenter.cpp:
```diff
@@ -154,23 +154,32 @@ BackgroundSubtraction ExperimentPresenter::backgroundSubtractionFromView() {
+  auto const polCorrOption = m_view->getPolarizationCorrectionOption();
 
-  if (!correctionsChecked) {
+  if (polarizationCorrectionTypeFromString(polCorrOption) == PolarizationCorrectionType::None) {
```
Thanks for this. There were a lot of little uncertainties and I was starting to get worried about some scope creep, so thought it was worth getting it checked before continuing with the tidying. There are a few things (such as the "Correction method is not defined" issue) that I left out to maintain consistency with the previous behaviour. But would be good to get fixed as part of this. I'll work through these and get it more tidied up.

Commit:
Allow selection of explicit None for flood workspace

RE #34720

Commit:
Stop defaulting to parameter file for empty workspace

This is to match up with the flood correction behaviour. Which does no
corrections if the workspace path or name is not given.

RE #34720

Commit:
Add a file handler to the exp presenter

RE #34720

Commit:
Add show valid controls to flood and pol corrs

RE #34720

Commit:
Show invalid pol corr file path on GUI

RE #34720

Commit:
Show invalid flood path on gui

RE #34720

Commit:
Correctly edit efficiencies on inst change

RE #34720

Commit:
Fix file path appearing with error when disabled.

RE #34720

Commit:
Add to the testing instructions

RE #34720

Commit:
Use auto & where appropriate

RE #34720

ConorMFinn commented:
> This is a minor thing and may just be my lack of understanding of how things are intended to work, but when I tried to load using a file name it required me to include the file extension for .nxs files but not for .nx5 files. Is this expected?

This seems to be something with how RROA takes and handles paths, so I think is out of scope for this PR.

rbauststfc reviewed (commented):
The recent code changes look good and the functionality feels much more polished and user friendly. It's great that the manual testing documentation is updated in this PR too.

Just one thing I noticed about the file path validation is that the reduction succeeds if you just enter the name of a flood/efficiency file located in a Mantid user directory, however the GUI displays this as invalid. To get the input to display as valid I needed to enter the full file path, even though this wasn't actually needed for the reduction to work. I think it's definitely the right thing that we don't require the user to enter the full file path if it's already in their user directories, so would it be possible for the validation to be consistent with this?

Commit:
Check for valid shortened file paths

Make use of the FileFinder singleton to find file paths. This means that
just giving the filename on its own should be enough to reach the file.
The algorithms underneath do this anyway, so this is mainly to make sure
that the validation on the GUI matches up with the valid paths for the
algortithms.

RE #34720

rbauststfc reviewed (approved):
The latest changes to provide accurate validation in the GUI when just a filename is used are working really nicely and the code for it also looks really good. Thanks very much for making the effort to get this last bit working! This all looks ready to go in now 👍
